[CodeGen] [ARM] Make RISC-V Init Undef Pass Target Independent and add support for the ARM Architecture.

Stylie777 opened the pull request:
When using Greedy Register Allocation, there are times where early-clobber values are ignored, and assigned the same register. This is illeagal behaviour for these intructions. To get around this, using Pseudo instructions for early-clobber registers gives them a definition and allows Greedy to assign them to a different register. This then meets the ARM Architecture Reference Manual and matches the defined behaviour.

This patch takes the existing RISC-V patch and makes it target independent, then adds support for the ARM Architecture. Doing this will ensure early-clobber restraints are followed when using the ARM Architecture. Making the pass target independent will also open up possibility that support other architectures can be added in the future.

jthackray commented on llvm/lib/Target/ARM/ARMInitUndef.cpp:
```diff
@@ -0,0 +1,242 @@
+// This pass runs to allow for Undef values to be given a definition in
+// early-clobber instructions. This can occur when using higher levels of
+// optimisations. Before, undef values were ignored and it would result in
+// early-clobber instructions using the same registeres for the output as one of
```
nit: typo s/registeres/registers/

Stylie777 replied:
Fixed this, although I think the comment has now been removed.

tmatheson-arm commented:
How similar is this to the RISC-V pass? Can it be made one generic pass that any backends could use? The problem does not seem specific to these two architectures.

Stylie777 commented:
> How similar is this to the RISC-V pass? Can it be made one generic pass that any backends could use? The problem does not seem specific to these two architectures.

@tmatheson-arm This would require some investigation into how this would work - I can look into this

Stylie777 commented:
@tmatheson-arm to answer your question, the passes are similar to one another in some aspects and different in others. The logic employed is very similar, with how it checks for early-clobber and undef values. The differences come in how it applies the Pseudo instruction. For this solution with ARM, we have one instruction which takes on the form of a `ARM::MQPRRegClass` type. RISC V has multiple different ways of allocating the pseudo instructions, their version uses four different instructions that can be used depending on the situation.

That difference is my concern in making this a generic pass, and how it would deal with the different architectures and register configurations.

ostannard commented:
To make this pass target-independent, you could move the target-specific bits (`getVRLargestSuperClass`, `isVectorRegClass`, `getUndefInitOpcode`, ...) into `TargetRegisterInfo` and `TargetInstrInfo`, which already have a lot of virtual functions for querying target-specific information from target-independent classes.

Stylie777 committed:
[CodeGen] Convert RISCV Init Undef pass to support any architecture

Currently this pass is designed for RISC-V only, however this is
not the only architecture with the bug reported in issue #50157.
We can convert the exisiting pass to be generic, using some of the
existing Parent classes rather than RISC-V specific classes to
bring the same functionality to other Architectures.

The pass has been refactored, removing the RISC-V specific functions
and data-types and replacing them with datatypes that will support
all architectures and virtual functions in the respecitive classes
that allow support for the pass to be added. By default, this pass
will not run on on all architectures, only those that have the
`supportsInitUndef` function impletmented. This commit will only
refactor the exisiting code and add the pass in as a CodeGen pass rather
than target specific. To add support for other architectures, this should
be done in new, following commits.

This will allow for the pass to be used by any architecture. With the
correct overriding functions, other architectures can be supported
to provide the same functionality as was added to fix issue that
was reported in Issue #50157.

This is still a temporary measure, and a better more permenant fix
should be found, but for the time being this will allow for the
correct early-clobber contraint to be followed when defined in
vector instructions.

Stylie777 commented:
I have updated this PR to now be a refactor of the RISC-V pass to make it target independent, followed by adding support for the ARM Architecture.

Stylie777 committed:
[ARM] Add support for ARM Instructions for Init Undef Pass

For assembly instructions within the ARM Architecture, certain
instructions have early-clobber restraints that are defined
for the instruction. However, when using the Greedy register
allocator this is ignored. To get around this, we can use
the Init Undef pass to assign a Pseudo instruction for the
registers that are early-clobber to ensure the restraint
is followed.

This adds in support for this using a new Pseudo instruction,
`PseudoARMInitUndef` which is used to ensure early-clobber
restrains are followed. The relevant overriding functions
have also been provided to ensure the architecture is
supported by the pass and the required information can be
passed to ensure that early-clobber restrains are respected.

arichardson commented on llvm/include/llvm/CodeGen/CodeGenPassBuilder.h:
```diff
@@ -1179,6 +1179,8 @@ void CodeGenPassBuilder<Derived>::addOptimizedRegAlloc(
     AddMachinePass &addPass) const {
   addPass(DetectDeadLanesPass());
 
+  addPass(InitUndefPass())
```
If suggest running git-clang-format.

Stylie777 replied:
I have already run this locally and will be updated in the next patch set

Stylie777 commented on llvm/lib/CodeGen/InitUndef.cpp:
```diff
@@ -152,7 +118,7 @@ bool RISCVInitUndef::handleReg(MachineInstr *MI) {
     if (!UseMO.getReg().isVirtual())
       continue;
-    if (!isVectorRegClass(UseMO.getReg()))
+    if (!TII->isVectorRegClass(MRI->getRegClass(UseMO.getReg())))
```
Actually, a second thought @ostannard, is this function needed? This makes sure there is a Pseudo instruction available but if one has not been implemented the issue will still exist for that instruction. 

I propose to remove the function and check all together, this will ensure that when adding support for new architectures, all the register classes that require support have the relevant pseudo instruction added.

ostannard replied:
Yes, I think it would be better to delete this check, as the register class shouldn't change whether that transformation is necessary or not. My main concern here is that there's actually some other condition, and the original RISC-V implementation was using this as a proxy for it which happened to be correct for RISC-V, but not in general. Removing the condition and running the tests seems like a good way to check that.

Stylie777 replied:
Check has been deleted. Tests pass locally so functionality is not lost and there are no new issues raised by removing the check.

Stylie777 commented on llvm/include/llvm/CodeGen/TargetInstrInfo.h:
```diff
@@ -2223,6 +2223,25 @@ class TargetInstrInfo : public MCInstrInfo {
     llvm_unreachable("unknown number of operands necessary");
   }
 
+  virtual unsigned getUndefInitOpcode(unsigned RegClassID) const {
```
Comments have been added to the functions that are added by this PR.

Stylie777 commented on llvm/include/llvm/CodeGen/TargetSubtargetInfo.h:
```diff
@@ -327,6 +327,15 @@ class TargetSubtargetInfo : public MCSubtargetInfo {
+  // This was originally used in RISC-V's Init Undef pass but has been moved to
+  // be a virtual function when the pass was refactored to support multiple
+  // architectures.
+  virtual bool hasVInstructions() const { return false; }
```
This function has been renamed to better reflect its functionality to `supportsInitUndef()`. The RISC V Implementation calls the `hasVInstructions()` function to functionality is the same for the architecture. 

I have also added a check for AdvSIMD instructions on Arm.

Stylie777 commented on llvm/lib/CodeGen/InitUndef.cpp:
```diff
@@ -7,22 +7,21 @@
-// allocator allocated same or overlapped if the input register is undef value,
-// so convert IMPLICIT_DEF to temporary pseudo instruction and remove it later
-// could prevent that happen, it's not best way to resolve this, and it might
+// Certain vector instructions have register overlapping constraints, and
```
The comment has been update to be more generic, rather than focusing on vector instructions.

ostannard commented on llvm/include/llvm/CodeGen/TargetSubtargetInfo.h:
```diff
@@ -327,6 +327,14 @@ class TargetSubtargetInfo : public MCSubtargetInfo {
+  /// the Init Undef Pass. By default, it is assumed that it will not support
+  /// the pass, with architecture specific overrides providing the information
+  /// where they are implemented. This was originally used in RISC-V's Init
+  /// Undef pass but has been moved to be a virtual function when the pass was
```
Nit: You don't need to talk about the history here, this should just describe the current state of the code.

ostannard commented on llvm/lib/Target/ARM/ARMBaseRegisterInfo.h:
```diff
@@ -240,6 +240,33 @@ class ARMBaseRegisterInfo : public ARMGenRegisterInfo {
+    if (ARM::DPR_VFP2RegClass.hasSubClassEq(RC))
+      return &ARM::DPR_VFP2RegClass;
+    if (ARM::GPRRegClass.hasSubClassEq(RC))
+      return &ARM::DPR_VFP2RegClass;
```
Copy-paste error?

ostannard commented on llvm/lib/Target/ARM/ARMSubtarget.h:
```diff
@@ -278,6 +278,14 @@ class ARMSubtarget : public ARMGenSubtargetInfo {
   }
 
+  /// Returns true as the ARM Architecture is supported by the Init Undef Pass.
+  /// We want to enable this for MVE and NEON instructions, however this can be
```
Are there any NEON instructions which this is needed for?

ostannard commented on llvm/test/CodeGen/Thumb2/mve-vmull-splat.ll:
```diff
@@ -190,12 +190,17 @@ entry:
 ; CHECK-LABEL: sext32_0213_0ext:
 ; CHECK:       @ %bb.0: @ %entry
+; CHECK-NEXT:    .vsave {d8, d9}
+; CHECK-NEXT:    vpush {d8, d9}
```
The generated code for these tests has gotten longer, was this previously generating invalid code, or is this an unwanted side-effect of this pass?

Stylie777 replied:
This is something that is due to the new Pass being enabled on Arm. This test, and the other test that has instructions added in this file require the function `handleSubReg` to be called, and such initialise the registers in a different way. This is causing extra instructions on the Arm Architecture, however as a limited number of test cases are effected I deem this an edge case and as such we should allow the compiler to generate the extra instructions.

ostannard replied:
Fair enough, if there's no easy fix then correctness needs to take priority over performance.

ostannard reviewed (approved):
LGTM, thanks!
